### PR TITLE
Fix(visual start) Make tasks in starting stage to appear parallel.

### PR DESCRIPTION
**Pull request by abhinandan13jan**
Fixes: https://jira.coreos.com/browse/ODC-1844
--------------------------------------------------------------------------
before fix:
Screenshot from 2019-09-12 19-40-06
--------------------------------------------------------------------------
post fix:
Screenshot from 2019-09-26 15-47-35
Screenshot from 2019-09-26 15-47-45
Screenshot from 2019-09-26 15-48-02
Screenshot from 2019-10-09 18-48-37

### CODE REVIEW

**Comment by invincibleJai** [comment]
@abhinandan13jan was verifying the changes, one observation does task name should be part of the tooltip as it used to be earlier?  

Before Fix: 

Screenshot 2019-09-16 at 6 09 45 PM

After Fix:

Screenshot 2019-09-16 at 6 10 05 PM

Task name `hello-world-truncate-more-than...` is not visible in tooltip postfix and issue doesn't have much info as even tooltip style seems changed.

**Comment by abhinandan13jan** [comment]
> @abhinandan13jan was verifying the changes, one observation does task name should be part of the tooltip as it used to be earlier?
> 
> Before Fix:
> 
> Screenshot 2019-09-16 at 6 09 45 PM
> 
> After Fix:
> 
> Screenshot 2019-09-16 at 6 10 05 PM
> 
> Task name `hello-world-truncate-more-than...` is not visible in tooltip postfix and issue doesn't have much info as even tooltip style seems changed.
> 
> Blank tooltip may as well appear as below
> 
> Screenshot 2019-09-16 at 6 22 49 PM
> 
> **NOTE**: Before fix screenshots are from cluster directly and after fix is from this PR. Tooltip doesn't seem changed in this PR but wanted to highlight it.

@invincibleJai I cant really figure out why you see changes in the tooltip area but that isn't something this PR should affect

**Comment by invincibleJai** [comment]
@abhinandan13jan yes you are right and sorry I got confused same I mentioned in note. Wanted to highlight tooltip behavior not sure if as per ask from UXD or an issue. 

> NOTE: Before fix screenshots are from cluster directly and after fix is from this PR. Tooltip doesn't seem changed in this PR but wanted to highlight it.

**Comment by karthikjeeyar** [comment]
@abhinandan13jan rebase with master and push again

**Comment by andrewballantyne** [comment]
There are some UI work still needed here. Maybe even some UX (cc @serenamarie125).

image

Parallel tasks are shown stacking vertically, but they seem to also implicitly expect there to be a node to the left and right. This is not the case for the `simple-pipeline`.

Note: We are likely seeing the lack of the horizontal line leaving the "hello-world-1" node because it originates from the following node - of which is not in this example.

Should we be looking at just the pills stacking without lines for this example? Please check with UX @abhinandan13jan

**Comment by andrewballantyne** [comment]
> Hey guys, this particular fix deals with changing the data structure only. If there are requirements to change the visual part/css stuff, we will have to look into it in a separate PR. I rebased the branch with master.

In my opinion, the visual issues here are not release worthy. Until we get feedback, I don't think we should look to merge this. Although somewhat confusing, the previous design does not look broken. I think we are better without this in than with this in as-is.

**Comment by christianvogt** [comment]
As @andrewballantyne points out, there are visual issues with the lines now. Although the current implementation shows the steps in series instead of parallel, it doesn't have visual flaws. The lines needs to be fixed here too.

**Comment by andrewballantyne** [comment]
@abhinandan13jan please target this PR to 4.3 and rebase as needed.

UX has suggested we go with no unnecessary lines when showing parallel tasks. So if no tasks are preceding, then no lines going in from the left; likewise, no tasks following up means no lines either. AKA only lines between pills and no orphaned lines or pills.

We should have 3 example states to test the style changes:

- A pipeline with just 3 parallel tasks (like the one you have in your screenshots)
- A pipeline with an initial solo task + at least 2 or more parallel tasks
- A pipeline with at least 2 parallel tasks followed up with a single task

**Comment by andrewballantyne** [comment]
@serenamarie125 feel free to add anything else if I omitted something or if you have follow-up comments.

File: frontend/packages/dev-console/src/components/pipelines/PipelineVisualizationGraph.scss [hunk 1]
```diff
@@ -36,30 +62,109 @@ $gutter: 1.7em;
...
+            right: -17px;
+            border-left: 1px solid var(--pf-global--BorderColor--light-100);
+            border-radius: 20px 0 0;
+          }
         }
-        &::after {
-          right: 0;
-          transform: translateX(100%);
+        //connect each task
+        &:not(:first-child) {
```
Comment (karthikjeeyar) [inline frontend/packages/dev-console/src/components/pipelines/PipelineVisualizationGraph.scss]: Next three blocks of code are duplicated on top of the file as well, we could avoid this using `extends`

**Comment by serenamarie125** [comment]
@abhinandan13jan when there is series of parallel tasks each grouped based on common "afterTask" identifiers, the desired UX would be to change the connectors a bit to include more space - similar to columns 2 & 3 in the mock below:
Screen Shot 2019-09-26 at 9 33 23 AM

File: frontend/packages/dev-console/src/components/pipelines/PipelineVisualizationGraph.scss [hunk 2]
```diff
@@ -36,30 +64,109 @@ $gutter: 1.7em;
...
+          &::after {
+            top: -1.9em;
+            height: 3.3em;
+          }
         }
       }
     }
+  }
+
+  &__stage {
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/pipelines/PipelineVisualizationGraph.scss]: You close and then re-open `&__stage` - recommend just removing these lines.

File: frontend/packages/dev-console/src/utils/__tests__/pipeline-utils.spec.ts [hunk 3]
```diff
@@ -10,9 +10,8 @@ import { mockPipelinesJSON } from './pipeline-test-data';
 describe('pipeline-utils ', () => {
   it('For first pipeline there should be 2 stages of length [0:[1],1:[2]]', () => {
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/utils/__tests__/pipeline-utils.spec.ts]: Noticed one other thing - the test name should change as well.

File: frontend/packages/dev-console/src/components/pipelines/PipelineVisualizationGraph.scss (hunk 2 above)
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelines/PipelineVisualizationGraph.scss]: updated

File: frontend/packages/dev-console/src/utils/__tests__/pipeline-utils.spec.ts (hunk 3 above)
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/utils/__tests__/pipeline-utils.spec.ts]: updated

**Comment by karthikjeeyar** [comment]
@abhinandan13jan The connectors are slightly off, apart from that code changes looks good to me

screenshot-localhost-9000-2019 10 14-20-20-44
screenshot-localhost-9000-2019 10 14-20-19-34

**Comment by abhinandan13jan** [comment]
@karthikjeeyar Updated

**Pull request merged**